### PR TITLE
Using Redis with sentinels #11617

### LINKED ISSUE #11617: Using Redis with sentinels
We use Redis distributed cache with 3 sentinels:
1Master + 1 sentinel in first PC
1Replica + 2 sentinels in second PC
Connection configuration:
```"OrchardCore": {
    "OrchardCore_Redis": {
      "Configuration": "localhost:26379,password=somePassword,serviceName=redis-orchard,allowAdmin=true", 
      "InstancePrefix": "orchard" // Optional prefix allowing a Redis instance to be shared by different applications.
    }
```
For using this configuration, we use main branch https://github.com/StackExchange/StackExchange.Redis (previous versions, include last release [2.5.61], throw error about getting information about replica).
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <RootNamespace>OrchardCore.Redis</RootNamespace>
    <!-- NuGet properties-->
    <Title>OrchardCore Redis Abstractions</Title>
    <Description>$(OCFrameworkDescription)

    Abstractions for Redis.</Description>
    <PackageTags>$(PackageTags) OrchardCoreFramework Abstractions</PackageTags>
  </PropertyGroup>

  <ItemGroup>
    <FrameworkReference Include="Microsoft.AspNetCore.App" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore.DataProtection.StackExchangeRedis" />
    <PackageReference Include="Microsoft.Extensions.Caching.StackExchangeRedis" Version="5.0.1" />
    <PackageReference Include="StackExchange.Redis" Version="2.6.34-gc83443aaaf" />
  </ItemGroup>

  <ItemGroup>
    <ProjectReference Include="..\OrchardCore.Abstractions\OrchardCore.Abstractions.csproj" />
  </ItemGroup>

</Project>
```
1. Configure OrchardCore_OrchardCore_Redis
2. Enable feature OrchardCore.Redis.Cache (feature OrchardCore.Redis is enabled as dependency)
3. Error is thrown: "Sentinel: The ConnectionMultiplexer is not a Sentinel connection. Detected as: Stanalone"

How it works in code:
1. OrchardCore.Redis.Startup ConfigureServices(IServiceCollection services) configures RedisOption from configuration["OrchardCore_Redis:Configuration"].
2. When OrchardCore.Redis is started OrchardCore.Redis.Services.RedisServer connects to Redis with ConnectAsync(_options.Value.ConfigurationOptions) (use configured in step 1)
3. When StackExchange.Redis connecting to localhost:26379 (sentinel service), ConfigurationOptions is changing because sentinel service reports about master and replica servers.
4. After connection ConfigurationOptions  is set as "[1.1.1.1:6379,2.2.2.2:6379],password=somePassword,serviceName=redis-orchard,allowAdmin=true". This action changes RedisOption.ConfigurationOptions  too.
5. When OrchardCore.Redis.Cache  is started  AddStackExchangeRedisCache use RedisCacheOptionsSetup for connection to Redis.
6. RedisCacheOptionsSetup uses RedisOption.ConfigurationOptions   (now it is [1.1.1.1:6379,2.2.2.2:6379],password=somePassword,serviceName=redis-orchard,allowAdmin=true) and when StackExchange.Redis start connection it get first EndPoint (1.1.1.1:6379)
7. Because ConfigurationOptions contains serviceName=redis-orchard, StackExchange.Redis defines connection how sentinel but ServerSelectionStrategy.ServerType (by EndPoint) is Standalone and error is thrown.

My propose:
RedisCacheOptions and RedisOption should be distinct.
1. OrchardCore.Redis.Options.RedisCacheOptionsSetup should be deleted.
2. RedisCacheStartup should configure RedisCacheOptions from configuretion.
```
    [Feature("OrchardCore.Redis.Cache")]
    public class RedisCacheStartup : StartupBase
    {
        private readonly string _tenant;
        private readonly IShellConfiguration _configuration;
        private readonly ILogger _logger;

        public RedisCacheStartup(ShellSettings shellSettings, IShellConfiguration configuration, ILogger<Startup> logger)
        {
            _tenant = shellSettings.Name;
            _configuration = configuration;
            _logger = logger;
        }

        public override void ConfigureServices(IServiceCollection services)
        {
            if (services.Any(d => d.ServiceType == typeof(IRedisService)))
            {
                try
                {
                    var configurationOptions = ConfigurationOptions.Parse(_configuration["OrchardCore_Redis:Configuration"]);
                    var instancePrefix = _configuration["OrchardCore_Redis:InstancePrefix"];

                    services.Configure<RedisCacheOptions>(options =>
                    {
                        options.ConfigurationOptions = configurationOptions;
                        options.InstanceName = instancePrefix;
                    });
                }
                catch (Exception e)
                {
                    _logger.LogError("'Redis.Cache' features are not active on tenant '{TenantName}' as the 'Configuration' string is missing or invalid: " + e.Message, _tenant);
                    return;
                }

                services.AddStackExchangeRedisCache(o => { });
                services.AddScoped<ITagCache, RedisTagCache>();
            }
        }
    }
```
In this case all works as expected.

### PR DESCRIPTION
**Pull request by mazuryv**
PR is created.

Fixes #11617

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/OrchardCMS/OrchardCore?pullRequest=11637) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/OrchardCMS/OrchardCore?pullRequest=11637) before we can accept your contribution.<br/><br/>:x: mazuryv [sign now](https://cla.dotnetfoundation.org/OrchardCMS/OrchardCore?pullRequest=11637)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/OrchardCMS/OrchardCore?pullRequest=11637) it.</sub>

**Comment by jtkech** [comment]
Maybe we could keep the same pattern in the startup, meaning still doing a parse in a try/catch but just to validate the config string once and before registering the services, so here the only update would be to set `options.Configuration` to `originalConfigurationString` that you can rename to `configurationString`.

Then, looks good to parse it in the ConfigurationOptions getter.

What you think about this comment https://github.com/dotnet/aspnetcore/issues/28367#issue-756609999 saying the following

> However, once the master fails and another master is elected by the sentinels, this connection object is no longer valid and the caching breaks. There is currently no way to re-instantiate the connection easily.

**Comment by mazuryv** [comment]
I check it:

> However, once the master fails and another master is elected by the sentinels, this connection object is no longer valid and the caching breaks. There is currently no way to re-instantiate the connection easily.
When sentinels elect new master from slave - looks like connection object is still valid and connect to another server without reconection.

**Comment by jtkech** [comment]
Okay cool, thanks for trying it.

> I think try/catch protected only against parse exception, not against bad configuration.

Yes but still useful to do as a first level check before registering the related services. But yes keep the change to set the Configuration string `options.Configuration = configurationString`.

> When public ConfigurationOptions ConfigurationOptions => ConfigurationOptions.Parse(Configuration); is used, we always get correct configuration.

Then yes, also keep this change.

Need to leave, will look at it again this night.

File: src/OrchardCore.Modules/OrchardCore.Redis/Startup.cs [hunk 1]
```diff
@@ -32,22 +32,20 @@ public Startup(ShellSettings shellSettings, IShellConfiguration configuration, I
...
-                {
-                    options.ConfigurationOptions = configurationOptions;
-                    options.InstancePrefix = instancePrefix;
-                });
-            }
-            catch (Exception e)
+            if (string.IsNullOrEmpty(originalConfigurationString))
             {
-                _logger.LogError("'Redis' features are not active on tenant '{TenantName}' as the 'Configuration' string is missing or invalid: " + e.Message, _tenant);
+                _logger.LogError("'Redis' features are not active on tenant '{TenantName}' as the 'Configuration' string is missing", _tenant);
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Redis/Startup.cs:40]: Can you keep the error message in a different param? That might help investigate failures

Comment (mazuryv) [reply src/OrchardCore.Modules/OrchardCore.Redis/Startup.cs]: Ok. I kept it.

File: src/OrchardCore/OrchardCore.Redis.Abstractions/RedisOptions.cs [hunk 2]
```diff
@@ -2,16 +2,16 @@
 
 namespace OrchardCore.Redis
 {
-    /// <summary>
-    /// Redis options.
-    /// </summary>
     public class RedisOptions
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.Redis.Abstractions/RedisOptions.cs:5]:
```suggestion
    /// <summary>
    /// Configuration options for <see cref="IRedisService"/>.
    /// </summary>
    public class RedisOptions
```

As done here https://github.com/dotnet/aspnetcore/blob/39219394f66854db5d556ad78f9bb065ee48b585/src/Caching/StackExchangeRedis/src/RedisCacheOptions.cs#L12-L15

File: src/OrchardCore/OrchardCore.Redis.Abstractions/RedisOptions.cs [hunk 3]
```diff
@@ -2,16 +2,16 @@
...
     {
         /// <summary>
         /// The configuration used to connect to Redis.
         /// </summary>
-        public ConfigurationOptions ConfigurationOptions { get; set; }
-
+        public ConfigurationOptions ConfigurationOptions => ConfigurationOptions.Parse(Configuration);
+        /// <summary>
+        /// Original configuration string used to connect to Redis
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.Redis.Abstractions/RedisOptions.cs:12]:
```suggestion
        /// The configuration string used to connect to Redis.
```

**Review by jtkech — APPROVED** [review]
@mazuryv I approved

Just 2 comment suggestions to be applied.

**Comment by jtkech** [comment]
@mazuryv I approved

Just 2 comment suggestions to be applied before merging

File: src/OrchardCore/OrchardCore.Redis.Abstractions/RedisOptions.cs [hunk 4]
```diff
@@ -3,15 +3,18 @@
...
         /// <summary>
         /// The configuration used to connect to Redis.
         /// </summary>
-        public ConfigurationOptions ConfigurationOptions { get; set; }
-
+        public ConfigurationOptions ConfigurationOptions => ConfigurationOptions.Parse(Configuration);
+        /// <summary>
+        /// The configuration string used to connect to Redis.
+        /// </summary>
+        public string Configuration { get; set; }
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.Redis.Abstractions/RedisOptions.cs:17]: Last thing, please add a new line between each property, I can't edit your file.

**Comment by jtkech** [comment]
@mazuryv Last thing, please add a new line between each RedisOptions property, I can't edit your file.

File: src/OrchardCore/OrchardCore.Redis.Abstractions/RedisOptions.cs (hunk 4 above)
Comment (jtkech) [inline src/OrchardCore/OrchardCore.Redis.Abstractions/RedisOptions.cs:17]:
```suggestion

        /// <summary>
        /// The configuration string used to connect to Redis.
        /// </summary>
        public string Configuration { get; set; }

```

**Comment by mazuryv** [comment]
Sorry, I was offline.
I prepare little cosmetic change if you want to approve it.

**Comment by jtkech** [comment]
No problem

> I prepare little cosmetic change if you want to approve it.

I already approved it.

Only need to wait for the next triage ;)

**Comment by jtkech** [comment]
There is no breaking change, we still generate `ConfigurationOptions` objects by parsing the `Configuration` string but lazily. The problem was that when connecting with sentinels the `ConfigurationOptions` is mutated, so e.g. after our service was connected, redis cache was trying to connect with a bad `ConfigurationOptions`, that we now generate lazily from the Configuration string.

**Pull request merged**